src/attr/_make.py

```diff
@@ -347,75 +511,39 @@ def attrs(maybe_cls=None, these=None, repr_ns=None,
     ..  versionchanged::
             17.1.0 *hash* supports ``None`` as value which is also the default
             now.
+    ..  versionchanged:: 17.3.0
+            All classes are replaced instead of attaching methods to the
+            original class.
     """
     def wrap(cls):
         if getattr(cls, "__class__", None) is None:
             raise TypeError("attrs only works with new-style classes.")
 
-        if repr is False and str is True:
-            raise ValueError(
-                "__str__ can only be generated if a __repr__ exists."
-            )
-
-        ca_list = _transform_attrs(cls, these)
+        builder = _ClassBuilder.from_class(cls, these, slots, frozen)
 
-        # Can't just re-use frozen name because Python's scoping. :(
-        # Can't compare function objects because Python 2 is terrible. :(
-        effectively_frozen = _has_frozen_superclass(cls) or frozen
         if repr is True:
-            cls = _add_repr(cls, ns=repr_ns, str=str)
+            builder.add_repr(repr_ns)
+        if str is True:
+            builder.add_str()
         if cmp is True:
-            cls = _add_cmp(cls)
+            builder.add_cmp()
 
         if hash is not True and hash is not False and hash is not None:
+            # Can't use `hash in` because 1 == True for example.
             raise TypeError(
                 "Invalid value for hash.  Must be True, False, or None."
             )
         elif hash is False or (hash is None and cmp is False):
-            pass
+            builder.add_hash_by_id()
```

Comment:
No, because the whole point of this patch is that dict classes will behave like slots classes now.  A comment should be added tho indeed!  (your question genuinely confused me and made me think it through one more time, so thanks again for that!)